### ISSUE TITLE
When overlapping sections occur, what should the user do about it?

### ISSUE THREAD

**Issue by sageserpent-open** [sageserpent-open/kineticMerge#46]
Saw this when using Kinetic Merge on its own repository, merging two branches to make Git commit SHA: e1554c515b0b66f9c56147bb06c832b85f2f11fe.

Just running Kinetic Merge to merge the two parent histories with its defaults yields overlapping sections; there are also plenty of these seen when running `CodeMotionAnalysisTest`, eg:

```
com.sageserpent.kineticmerge.core.MappedContentSources$OverlappingSections: Overlapping section detected on side: left at path: 8, Section(label = "left", path = 8, startOffset = 79, size = 5, content = Vector(2, 5, 1, 1, 3)) (content: Vector(2, 5, 1, 1, 3)) overlaps with start of section: Section(label = "left", path = 8, startOffset = 82, size = 5, content = Vector(1, 3, 8, 3, 4)) (content: Vector(1, 3, 8, 3, 4)), overlap content: Vector(1, 3).
```

Now, the diagnostic issued when running the full Kinetic Merge application renders the sections in the nicer way that reflects the underlying text, but either way, the end user is presented with a message that says **"It failed - over to you..."**.

Could we provide some intelligent hint about how to fix this?

I ended up increasing the minimum match size to workaround the overlaps and it worked - perhaps a minimum size hint can be provided? Not sure if it is worth trying to deduce whether this would guarantee lack of overlaps, especially if there might be several lots of overlaps, possibly at different match sizes...

**Comment by sageserpent-open** [comment]
While we're at it, there is an invariant set up by `CodeMotionAnalysis.of` that sections may only overlap if they share the same match size.

The overlap detection knows nothing about this and will report the first overlap it sees, which might not be the one with the largest match size (we want to fix that one as a priority if we increase the minimum match size).

Perhaps `CodeMotionAnalysis.of` should be draconian and throw out all matches that refer to at least one overlapping section. Worth a spike to see whether this can be delivered without killing performance and what the resulting merges would look like. If this looks good, this ticket can be changed over in favour of that...

**Comment by sageserpent-open** [comment]
Another example merge to demonstrate the problem is when merging in the Kinetic Merge respository from commit 48c967e4127e7a0d4034344e8e97c8139c565e58 to 9f43d7095bf4aaaad8e8d9b1152ba8098c34e436:

```bash
git checkout 9f43d7095bf4aaaad8e8d9b1152ba8098c34e436

kinetic-merge --no-commit 48c967e4127e7a0d4034344e8e97c8139c565e58
```

Thanks to commit 8464ce39d29573866fce8e26f98f96119a3b4de7 (sneaked into #35), the error message isn't too bad:

```scala
com.sageserpent.kineticmerge.core.MappedContentSources$OverlappingSections: Overlapping section detected on side: BASE: 7064dbab at path: .../src/main/scala/com/sageserpent/kineticmerge/core/CodeMotionAnalysis.scala, Section(
  label = "BASE: 7064dbab",
  path = .../src/main/scala/com/sageserpent/kineticmerge/core/CodeMotionAnalysis.scala,
  start = TextPosition(line = 1007, characterOffset = 17),
  onePastEnd = TextPosition(line = 1007, characterOffset = 51),
  startTokenIndex = 5287,
  sizeInTokens = 5,
  content = "flatMap { case CatsInclusiveRange("
) (content: Vector(WithTrailingWhitespace(Significant(flatMap),Whitespace( )), WithTrailingWhitespace(Significant({),Whitespace( )), WithTrailingWhitespace(Significant(case),Whitespace( )), Significant(CatsInclusiveRange), Significant(())) 
overlaps with start of section: Section(
  label = "BASE: 7064dbab",
  path = .../src/main/scala/com/sageserpent/kineticmerge/core/CodeMotionAnalysis.scala,
  start = TextPosition(line = 1007, characterOffset = 32),
  onePastEnd = TextPosition(line = 1007, characterOffset = 61),
  startTokenIndex = 5290,
  sizeInTokens = 5,
  content = "CatsInclusiveRange(start, end"
) (content: Vector(Significant(CatsInclusiveRange), Significant((), Significant(start), WithTrailingWhitespace(Significant(,),Whitespace( )), Significant(end))), overlap content: Vector(Significant(CatsInclusiveRange), Significant(()). 
Consider setting the command line parameter `--minimum-match-size` to something larger than 5.
```

Heeding the advice and using:

```bash
kinetic-merge --no-commit --minimum-match-size=6 48c967e4127e7a0d4034344e8e97c8139c565e58
```
yields a successful merge.

This is good, but could it be improved?

In the context of `CodeMotionAnalysis.of`, the only sections that can overlap are of the same size - it is possible to have unrelated sets of overlapping sections at different sizes, though. Perhaps these could be coalesced into single gap-filler sections and the corresponding matches completely or partially suppressed?

**Comment by sageserpent-open** [comment]
Doing a spike with this patch yields a successful (and reasonable) merge for the second example:

```text
Index: src/test/scala/com/sageserpent/kineticmerge/core/CodeMotionAnalysisTest.scala
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/test/scala/com/sageserpent/kineticmerge/core/CodeMotionAnalysisTest.scala b/src/test/scala/com/sageserpent/kineticmerge/core/CodeMotionAnalysisTest.scala
--- a/src/test/scala/com/sageserpent/kineticmerge/core/CodeMotionAnalysisTest.scala	(revision c24349097ddcac9f30079b5cd3f5919fb08c2588)
+++ b/src/test/scala/com/sageserpent/kineticmerge/core/CodeMotionAnalysisTest.scala	(date 1734882509647)
@@ -452,7 +452,14 @@
           baseSources,
           leftSources,
           rightSources
-        )(configuration) match
+        )(
+          configuration,
+          // NOTE: the test cases can exhibit matches with overlapping sections
+          // that intrude on the content the test is checking, so rather than
+          // quietly suppressing the matches, we let admissible failures for
+          // overlapping sections occur and reject the test case.
+          suppressMatchesInvolvingOverlappingSections = false
+        ) match
           case Right(analysis) =>
             // Check that all matches are consistent with the base sections...
             analysis.base.values.flatMap(_.sections).foreach { baseSection =>
Index: src/main/scala/com/sageserpent/kineticmerge/core/CodeMotionAnalysis.scala
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/scala/com/sageserpent/kineticmerge/core/CodeMotionAnalysis.scala b/src/main/scala/com/sageserpent/kineticmerge/core/CodeMotionAnalysis.scala
--- a/src/main/scala/com/sageserpent/kineticmerge/core/CodeMotionAnalysis.scala	(revision c24349097ddcac9f30079b5cd3f5919fb08c2588)
+++ b/src/main/scala/com/sageserpent/kineticmerge/core/CodeMotionAnalysis.scala	(date 1734882509637)
@@ -66,7 +66,8 @@
       leftSources: Sources[Path, Element],
       rightSources: Sources[Path, Element]
   )(
-      configuration: Configuration
+      configuration: Configuration,
+      suppressMatchesInvolvingOverlappingSections: Boolean = true
   )(using
       hashFunction: HashFunction
   ): Either[Throwable, CodeMotionAnalysis[Path, Element]] =
@@ -1420,6 +1421,41 @@
         withoutTheseMatches(redundantMatches) -> usefulMatches
       end withoutRedundantPairwiseMatchesIn
 
+      def purgedOfMatchesWithOverlappingSections(
+          enabled: Boolean
+      ): MatchesAndTheirSections =
+        if enabled then
+          def overlapsWithSomethingElse(aMatch: GenericMatch): Boolean =
+            // NOTE: the invariant already guarantees that nothing will be
+            // subsumed by the match's sections, so this is only testing for
+            // overlaps.
+            aMatch match
+              case Match.AllSides(baseSection, leftSection, rightSection) =>
+                overlapsOrIsSubsumedByOnBase(baseSection) ||
+                overlapsOrIsSubsumedByOnLeft(
+                  leftSection
+                ) || overlapsOrIsSubsumedByOnRight(rightSection)
+              case Match.BaseAndLeft(baseSection, leftSection) =>
+                overlapsOrIsSubsumedByOnBase(baseSection) ||
+                overlapsOrIsSubsumedByOnLeft(
+                  leftSection
+                )
+              case Match.BaseAndRight(baseSection, rightSection) =>
+                overlapsOrIsSubsumedByOnBase(
+                  baseSection
+                ) || overlapsOrIsSubsumedByOnRight(rightSection)
+              case Match.LeftAndRight(leftSection, rightSection) =>
+                overlapsOrIsSubsumedByOnLeft(
+                  leftSection
+                ) || overlapsOrIsSubsumedByOnRight(rightSection)
+
+          val overlapping =
+            sectionsAndTheirMatches.values.filter(overlapsWithSomethingElse)
+
+          withoutTheseMatches(overlapping)
+        else this
+      end purgedOfMatchesWithOverlappingSections
+
       @tailrec
       private def eatIntoLargerPairwiseMatchesUntilStabilized(windowSize: Int)(
           matches: collection.Set[GenericMatch],
@@ -1925,12 +1961,14 @@
       val withAllMatchesOfAtLeastTheSureFireWindowSize =
         MatchesAndTheirSections.withAllMatchesOfAtLeastTheSureFireWindowSize()
 
-      if minimumSureFireWindowSizeAcrossAllFilesOverAllSides > minimumWindowSizeAcrossAllFilesOverAllSides
-      then
-        withAllMatchesOfAtLeastTheSureFireWindowSize
-          .withAllSmallFryMatches()
-      else withAllMatchesOfAtLeastTheSureFireWindowSize
-      end if
+      (if minimumSureFireWindowSizeAcrossAllFilesOverAllSides > minimumWindowSizeAcrossAllFilesOverAllSides
+       then
+         withAllMatchesOfAtLeastTheSureFireWindowSize
+           .withAllSmallFryMatches()
+       else withAllMatchesOfAtLeastTheSureFireWindowSize)
+        .purgedOfMatchesWithOverlappingSections(
+          suppressMatchesInvolvingOverlappingSections
+        )
     end matchesAndTheirSections
 
     matchesAndTheirSections.checkInvariant()
```

**Comment by sageserpent-open** [comment]
To explore - should the matches with sections involved in overlaps be quietly partitioned off in the `CodeMotionAnalysis` result? That might allow tests that are sensitive to overlaps to examine them, rather than the current approahc of just rejecting the test case outright.

Also, it's worth trying a spike to see if the optimal match size estimation can be done without having to build a throwaway `MatchesAndTheirSections`. Wouldn't sorting groups of sections by side and path on their start offsets be enough?

**Comment by sageserpent-open** [comment]
Got something going (and tested) in 0cf5d80183117b9f840f09e9bf894574a3720e30.

This exposed some rather cavalier treatment of section equality that has been latent in the implementation for some time, since: ee8888eeb7c0c82eefa91ce833e7d5cd676ca709. This was ostensibly about:
```text
Tweak how content is rendered for sections produced by `MappedContentSourcesOfTokens`.
```

The change of the class `SectionImplementation` from a case-class to an ordinary class caused equality and hashing to take on identity semantics; this has been implicitly relied on by the implementation of `MatchesAndTheirSection` to ensure that newly-found all-sides matches of the same size as existing pairwise matches (generated by previous larger all-sides matches eating into bigger pairwise matches) were able to eat into the now unwanted pairwise matches, thus completely replacing them.

That approach broke `CodeMotionAnalysisTest.matchesWithOverlappingSections` added as part of this ticket, so commit 5217daf4d660d523a29a2826f2b44b13959e5b88 was back-patched in to go back to having value-based equality and hashing for `SectionImplementation`.

This worked, but there is some very questionable use of *identity* equality in subsumption checks that is required by the new value-based approach! Need to sort this out properly...

**Comment by sageserpent-open** [comment]
Another TODO: `MatchesAndTheirSections.estimateOptimalMatchSize` builds up a throwaway instance of `MatchesAndTheirSections` purely to get hold of runs of sections on the same part and on the same side, ordered by their start index.

Wouldn't it be easier to do this by getting all the sections for each side, grouping them by path and then sorting each group with `sortBy` or whatever? That might avoid yet more hammering of the fingertree implementation.

**Comment by sageserpent-open** [comment]
Nope, one more TODO: for some reason, the same section is being added in multiple times in the same fingertree instance even after `MatchesAndTheirSections` has undergone cleanup. This is connected to the stabilization logic that together with its caller `MatchesAndTheirSection.withMatches` ends up adding the same all-sides matches both within stabilization and in the caller, namely the ones that ate into pairwise matches.

**Comment by sageserpent-open** [comment]
Merged on to `main` in 0cdc09e29743ceac2519a7b47ffa9a750b8cfc96.

The example merge completes with a tidy merge, detecting two lots of divergent moves and some ambiguous ones.

**Comment by sageserpent-open** [comment]
This went out in Release 1.4.0.
